### ISSUE TITLE
Quantum Espresso Geometry Optimization Error Code 3 

### ISSUE THREAD

**Issue by ghost** [libAtoms/ExPyRe#22]
Hi all,
when running **geometry optimization with Quantum Espresso** via the calculation = 'relax' command with a small maximum number of relaxation steps (e.g. 20) the calculation will result in an error code 3.
The QE output still terminates with 
```
=------------------------------------------------------------------------------=
   JOB DONE.
=------------------------------------------------------------------------------=
```
as it should and includes all necessary information (i.e. energy and forces).
However, wfl will through an error and stop the iterative training.

_Is it possible to prohibit certain error messages to halt the whole program?_ 

Here is an example [QE_run.tar.gz](https://github.com/libAtoms/ExPyRe/files/9055006/QE_run.tar.gz)

**Comment by bernstei** [comment]
I agree that this needs at least better documentation, and quite probably just more thinking about how to deal with this, but this is why it's complex: that initial issue is that the espresso executable returns an error code, and the ASE interface decided that this means that it should raise a python exception. Then the pool parallelization running inside the remote job has to decide how to deal with this error.  Then the remote job wrapper has to deal with the remote error. 

Let me look over everything it's doing, make some suggestions for your particular case, and think about whether anythin can be changed to make the overall error handling more intuitive.

**Comment by bernstei** [comment]
If you know how to do it, I'd suggest setting up a little python script using the underlying ASE QE calculator, no expyre or wfl, to confirm that the ASE interface is even returning any useful data.  Just set up a small system that you can run in serial and a tiny max steps. If it does return data but gives an error, then we can think about how to get expyre/wfl to handle that cleanly.

**Comment by daisiduke** [comment]
I went through the ASE documentation and realized that ASE is doing optimization not like I was assuming. Instead of running one QE job with calculation='relax', the relaxation is done via the ase.optimize (e.g. applying BFGS) module stringing together several QE jobs with calculation='scf'. 
As far as I know this is not yet integrated in the wfl/calculator/espresso.py package. Unfortunately I will be on vacation until Sunday and can not address this topic directly. I will start drafting some code changes by next Monday.
For now, I have created a (very) simple example running a QE relaxation within ase of a water molecule via the run.py script below to give a first impression and maybe get feedback whether this should be included.
All files can be found here (without pw.x file - path has to be specified has to be provided):
[ase_geo_opt.tar.gz](https://github.com/libAtoms/ExPyRe/files/9056114/ase_geo_opt.tar.gz)

```
 import ase, ase.io, os
from ase.calculators.espresso import Espresso
from ase.optimize import BFGS
import pickle

## Change to actual path
qe_binary = '/home/daisy/code/qe-7.0/bin/pw.x'
if not os.path.exists(qe_binary):
    raise FileNotFoundError('QE binary %s does not exist'%qe_binary)

at = ase.io.read('water.xyz')
with open('QE_settings.pkl', 'rb') as handle:
    kwargs_this_calc = pickle.load(handle)
kwargs_this_calc['command'] =  ' mpirun -np 4 %s -in PREFIX.pwi > PREFIX.pwo'%qe_binary

run_dir = './'
all_changes = ['positions', 'numbers', 'cell', 'pbc', 'initial_charges', 'initial_magmoms']
properties_use = ['energy','forces']

at.calc = Espresso(directory=run_dir, **kwargs_this_calc)
opt = BFGS(at)
opt.run(fmax=0.001,steps=4)

ase.io.write('final.xyz',at)
```

**Comment by daisiduke** [comment]
I will do some testing on Monday :muscle:

**Comment by gelzinyte** [comment]
I think the way to do DFT relaxations is to use the DFT code itself, not ASE optimiser with DFT single point evaluations. It's much more efficient. For example, for one or two examples I tested ASE-LBFGS & ORCA-single-point took over 100 steps, whereas ORCA itself took 20 steps and more efficient ones, because it re-uses some things from previous optimisation steps. 

Relaxing structures with ORCA directly is implemented in [workflow's ORCA calculator](https://github.com/libAtoms/workflow/blob/main/wfl/calculators/orca/__init__.py). If there's enough interest in implementing that for QE, it might be useful to look at that? 

The gist: set `task="opt"` when initialising the `wfl.calculators.orca.ORCA` calculator (instead of default `task="engrad"`, which corresponds to just single-point evaluations). Then the input file is written with `Opt` keyword in the input file, which is all that is needed for ORCA to relax geometries. Once ORCA itself is done, `wfl.calculators.orca.ORCA` reads in the relaxed positions from output file to `wfl.calculators.orca.ORCA.extra_results` dictionary in addition to the usual relaxed energies and forces. If workflow's ORCA calculator is called via the generic calculator, relaxed atoms' positions are set to the returned atoms' objects.

**Comment by gabor1** [comment]
Orca could be doing wavefunction extrapolation - could that possibly change the number of steps though? Maybe it's using force field preconditioning. We have that in ASE as well but it's a bit of a hack to set it up

**Comment by bernstei** [comment]
We could wrap https://databases.fysik.dtu.dk/ase/ase/optimize.html#pyberny, if it's useful

**Comment by gabor1** [comment]
I think our FF preconditioner should be similar

— Gábor

Gabor Csanyi
Professor of Molecular Modelling
Engineering Laboratory
University of Cambridge

**Comment by bernstei** [comment]
It might. I wonder if the current heuristic for turning it off for small systems is too conservative.
